shell/common/shell.cc

```diff
@@ -1205,6 +1205,12 @@ void Shell::OnFrameRasterized(const FrameTiming& timing) {
   }
 }
 
+void Shell::OnCompositorEndFrame(size_t freed_hint) {
+  if (engine_) {
```

Comment:
Shell APIs are only safely accessible on the platform task runner. The engine OTOH can only be accessed on the UI task runner.  I believe you are on the raster task runner here. So this this API is not resilient to death of the shell, accesses the engine in a thread unsafe manner and is a non-standard shell method to boot.

Reply:
Its fine for some methods to have different threading restrictions. But when they veer from convention, please document the same and also add an FML_DCHECK. In this case though, it just not thread-safe so lets fix that first.

Reply:
Hmm. I want the shell to be the delegate here. This was originally living on the raster task runner.

I should be ok as long as I just jump to the UI task runner for this block with a weak ptr to the engine right?

Do I need to get the weak ptr on the platform task runner?